src/group/mod.rs

```diff
@@ -28,85 +31,172 @@ use std::collections::HashMap;
+    leader_group_id: u64,
 }
 
-impl Groups {
-    /// Create new Groups with given configuration
-    pub fn new(config: Vec<(u64, Vec<u64>)>) -> Self {
+impl<'a> Groups {
+    /// Create new Groups with given configuration.
+    pub(crate) fn new(config: Vec<(u64, Vec<u64>)>) -> Self {
         let mut indexes = HashMap::new();
+        let mut unresolved = Vec::new();
         for (group_id, members) in config {
             for id in members {
-                indexes.insert(id, group_id);
+                indexes.insert(id, (group_id, INVALID_ID));
+                unresolved.push(id);
             }
         }
 
         Self {
             indexes,
+            unresolved,
             ..Default::default()
         }
     }
 
-    /// Get group members by group id.
-    pub fn get_members(&self, group_id: u64) -> Vec<u64> {
-        self.indexes
-            .iter()
-            .filter_map(
-                |(node, gid)| {
-                    if *gid == group_id {
-                        Some(*node)
-                    } else {
-                        None
-                    }
-                },
-            )
-            .collect::<Vec<_>>()
+    pub(crate) fn set_leader_group_id(&mut self, leader_group: u64) {
+        self.leader_group_id = leader_group;
     }
 
-    /// Get group id by member id
-    pub fn get_group_id(&self, member: u64) -> Option<u64> {
-        self.indexes.get(&member).cloned()
+    /// Get group id by member id.
+    pub(crate) fn get_group_id(&self, member: u64) -> Option<u64> {
+        self.indexes.get(&member).map(|(gid, _)| *gid)
     }
 
-    /// Set delegate for a group. The delegate must be in the group.
-    pub fn set_delegate(&mut self, group_id: u64, delegate: u64) {
-        self.delegate_cache.insert(group_id, delegate);
+    /// Get a delegate for `to`. The return value could be `to` itself.
+    pub fn get_delegate(&self, to: u64) -> u64 {
+        match self.indexes.get(&to) {
+            Some((_, delegate)) => *delegate,
+            None => INVALID_ID,
+        }
+    }
+
+    // Pick a delegate for the given peer.
+    //
+    // The delegate must satisfy conditions below:
+    // 1. The progress state should be `ProgressState::Replicate`;
+    // 2. The progress has biggest `match`;
+    // If all the members are requiring snapshots, use given `to`.
+    fn pick_delegate(&mut self, to: u64, prs: &ProgressSet) {
+        let group_id = match self.indexes.get(&to) {
+            Some((_, delegate)) if *delegate != INVALID_ID => return,
+            Some((gid, _)) if *gid == self.leader_group_id => return,
+            Some((gid, _)) => *gid,
+            None => return,
+        };
+
+        println!("pick delegate for {}", to);
+        let (mut chosen, mut matched, mut bcast_targets) = (INVALID_ID, 0, vec![]);
+        for id in self.candidate_delegates(group_id) {
+            let pr = prs.get(id).unwrap();
+            if matched < pr.matched {
```

Comment:
No checking for the `ProgressState` ?

Reply:
No, because both all of them can be delegate in the current implementation.